homeassistant/components/ring/config_flow.py

```diff
@@ -39,48 +47,95 @@ class RingConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
-
-                return self.async_create_entry(
-                    title=user_input["username"],
-                    data={"username": user_input["username"], "token": token},
-                )
             except Require2FA:
                 self.user_pass = user_input
 
                 return await self.async_step_2fa()
-
             except InvalidAuth:
                 errors["base"] = "invalid_auth"
             except Exception:  # pylint: disable=broad-except
                 _LOGGER.exception("Unexpected exception")
                 errors["base"] = "unknown"
+            else:
+                await self.async_set_unique_id(user_input["username"])
+                return self.async_create_entry(
+                    title=user_input["username"],
+                    data={"username": user_input["username"], "token": token},
+                )
 
         return self.async_show_form(
-            step_id="user",
-            data_schema=vol.Schema(
-                {vol.Required("username"): str, vol.Required("password"): str}
-            ),
-            errors=errors,
+            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
         )
 
     async def async_step_2fa(self, user_input=None):
         """Handle 2fa step."""
         if user_input:
+            if self.reauth_entry:
+                return await self.async_step_reauth_confirm(
+                    {**self.user_pass, **user_input}
+                )
+
             return await self.async_step_user({**self.user_pass, **user_input})
 
         return self.async_show_form(
             step_id="2fa",
             data_schema=vol.Schema({vol.Required("2fa"): str}),
         )
 
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Handle reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Dialog that informs the user that reauth is required."""
+        errors = {}
+        assert self.reauth_entry is not None
+
+        if user_input:
+            user_input[CONF_USERNAME] = self.reauth_entry.data[CONF_USERNAME]
+            try:
+                token = await validate_input(self.hass, user_input)
+            except Require2FA:
+                self.user_pass = user_input
+                return await self.async_step_2fa()
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                data = {
+                    CONF_USERNAME: user_input[CONF_USERNAME],
+                    "token": token,
```

Comment:
It's good to use a constant for strings that are used more than once. I'd use a constant `CONF_TOKEN` for `"token"`.

Reply:
Hi Martin, yes agreed. I started to introduce constants with this PR and thought about replacing all the existing instances of "username", "password" and "token" in the integration then decided to do that in a separate PR.  I should have brought the CONF_TOKEN in at this stage but will definitely get to it. (I'm hoping to get the integration to platinum level soon and I know this is a base requirement :)

Reply:
Good, thanks!